### PR TITLE
VAULT-9427: Add read support to `sys/loggers` endpoints

### PR DESCRIPTION
**Pull request by ccapurso**
PR https://github.com/hashicorp/vault/pull/16111 added support to modify server logging verbosity both globally and for individual loggers. It did not, however, add the ability to read the current state of the logger verbosity levels. This PR introduces read handlers for the `sys/loggers` and `sys/loggers/:name` endpoints.

Additional more low-level changes:
- Modified `Core.SetLogLevelByName` to account for the potential of `Core.allLoggers` including loggers that share the same name
- Modified existing `sys/loggers` handlers to use `ParseLogLevel` from the logging helper package to remove redundant code
- Added checks to ensure `sys/loggers` endpoints do not handle base logger (name == `""`)
- Added a `TranslateLoggerLevel` func to the logging helper package to determine the logging verbosity of an `hclog.Logger`

Example output on a dev server (`vault server -dev`):

```
# Set policy logger to trace level
❯ curl -XPUT -H "X-Vault-Token: $VAULT_TOKEN" -d '{"level":"trace"}' "$VAULT_ADDR/v1/sys/loggers/policy"

# Fetch verbosity level of all loggers
❯ curl -s -XGET -H "X-Vault-Token: $VAULT_TOKEN" "$VAULT_ADDR/v1/sys/loggers" | jq .data
{
  "activity": "info",
  "audit": "info",
  "auth.token.auth_token_cf840257": "info",
  "core": "info",
  "core.cluster-listener": "info",
  "core.cluster-listener.tcp": "info",
  "core.router": "info",
  "expiration": "info",
  "expiration.job-manager": "info",
  "grpclogfaker": "info",
  "identity": "info",
  "identity.storagepacker.entities": "info",
  "identity.storagepacker.groups": "info",
  "identity.storagepacker.local-aliases": "info",
  "mfa": "info",
  "policy": "trace",
  "quotas": "info",
  "rollback": "info",
  "seal.shamir": "info",
  "secrets.cubbyhole.cubbyhole_079cd5ab": "info",
  "secrets.identity.identity_7b1c0163": "info",
  "secrets.kv.kv_efb8658b": "info",
  "secrets.system.system_3ae59f37": "info",
  "storage.cache": "info",
  "storage.inmem": "info",
  "storage.sealunwrapper": "info",
  "system": "info",
  "token": "info"
}

# Get verbosity level of just the policy logger
❯ curl -s -XGET -H "X-Vault-Token: $VAULT_TOKEN" "$VAULT_ADDR/v1/sys/loggers/policy" | jq .data
{
  "policy": "trace"
}

# Revert verbosity level of just the policy logger to what is provided in config
❯ curl -s -XDELETE -H "X-Vault-Token: $VAULT_TOKEN" "$VAULT_ADDR/v1/sys/loggers/policy"

# Get verbosity level of just the policy logger
❯ curl -s -XGET -H "X-Vault-Token: $VAULT_TOKEN" "$VAULT_ADDR/v1/sys/loggers/policy" | jq .data
{
  "policy": "info"
}
```

### CODE REVIEW

File: vault/core.go [hunk 1]
```diff
@@ -2868,17 +2869,22 @@ func (c *Core) SetLogLevel(level log.Level) {
 	}
 }
 
-func (c *Core) SetLogLevelByName(name string, level log.Level) error {
+// SetLogLevelByName sets the logging level of named logger to level provided
```
Comment (hghaf099) [inline vault/core.go:2872]: This is a great godoc. Thanks for adding those.

File: vault/logical_system.go [hunk 2]
```diff
@@ -4921,14 +4979,15 @@ func (b *SystemBackend) handleLoggersByNameDelete(ctx context.Context, req *logi
...
-	level, err := getLogLevel(b.Core.logLevel)
+	level, err := logging.ParseLogLevel(b.Core.logLevel)
 	if err != nil {
 		return logical.ErrorResponse(fmt.Sprintf("log level from config is invalid: %s", err.Error())), nil
 	}
 
-	err = b.Core.SetLogLevelByName(nameRaw.(string), level)
-	if err != nil {
-		return logical.ErrorResponse(fmt.Sprintf("invalid params: %s", err.Error())), nil
+	name := nameRaw.(string)
```
Comment (hghaf099) [inline vault/logical_system.go:4987]: nit: Do we need to check if name is an empty string here the same as above functions?

**Review by hghaf099 — APPROVED** [review]
Just a nit. Otherwise, looks good to me!

File: vault/core.go (hunk 1 above)
Comment (ccapurso) [reply vault/core.go:2872]: Of course, thanks for the feedback :)

File: vault/logical_system.go (hunk 2 above)
Comment (ccapurso) [reply vault/logical_system.go:4987]: Yes, we should. Thanks!

File: helper/logging/logger.go [hunk 3]
```diff
@@ -112,6 +112,8 @@ func ParseLogFormat(format string) (LogFormat, error) {
 	}
 }
 
+// ParseLogLevel returns the hclog.Level that corresponds with the provided level string.
+// This differs hclog.LevelFromString in that it supports additional level strings.
```
Comment (peteski22) [inline helper/logging/logger.go:116]: I think I took this code and refactored it into a function when adding the log file feature, so this is my bad and thanks for adding comments!

However, I think that the func should probably *not* include `notice` or `warning` as that hasn't (to my knowledge) been documented on our site. 

There's also differing info depending on whether you look at the docs for the [`server` command](https://developer.hashicorp.com/vault/docs/commands/server#log-level) or [Vault configuration](https://developer.hashicorp.com/vault/docs/configuration#log_level). 

`Trace, Debug, Error, Warn, Info.` vs `trace, debug, info, warn, *err*`.

```suggestion
// ParseLogLevel differs hclog.LevelFromString in that it supports fewer log level strings.
```

[My PR](https://github.com/hashicorp/vault/pull/18031) I have open at the moment to add rotation to log files has updates to the docs to make them all standardised, and I will also update `ParseLogLevel` to not handle `notice`, `warning`  or `""` which defaults to `info`.

File: vault/logical_system_test.go [hunk 4]
```diff
@@ -4772,66 +4772,60 @@ func TestProcessLimit(t *testing.T) {
...
-
-	return false
-}
-
 func TestSystemBackend_Loggers(t *testing.T) {
 	testCases := []struct {
-		level       string
-		expectError bool
+		level         string
+		expectedLevel string
```
Comment (peteski22) [inline vault/logical_system_test.go:4778]: I'll also fix these tests in my PR (and ensure it's merged after yours), to ditch the `notice` `warning` `err`, `""`. :(

File: vault/logical_system_test.go [hunk 5]
```diff
@@ -4894,78 +4951,91 @@ func TestSystemBackend_LoggersByName(t *testing.T) {
 	testCases := []struct {
 		logger            string
 		level             string
+		expectedLevel     string
```
Comment (peteski22) [inline vault/logical_system_test.go:4954]: ... and I'll fix these tests too (sorry!).

**Review by peteski22 — COMMENTED** [review]
LGTM, thanks for adding some of the docs etc. much appreciated.

**Pull request merged**